File: 75Review/BinarySearch/875.cpp
#include <vector>
using namespace std;

class Solution {
  public:
    int minEatingSpeed(vector<int> &piles, int h) {
        long long maxVal = 0;
        for (int i = 0; i < piles.size(); i++) {
            maxVal = max(maxVal, (long long)piles[i]);
        }

        // binary search this
        long long minEat = maxVal;
        long long l = 1, r = maxVal;
        while (l <= r) {
            long long m = l + (r - l) / 2;
            long long temp = 0;
            for (int &p : piles) {
                temp += (p / m);
                if (p % m != 0) temp++;
            }
            if (temp <= h) {
                minEat = m;
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        return minEat;
    }
};

File: src/2025/rusting/875.rs
struct Solution;

/*
The idea behind this is just binary search the posible solutions
*/
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut l: i64 = 1;
        let mut r: i64 = *piles.iter().max().unwrap() as i64;
        let mut res = r;

        while l <= r {
            let m = l + (r - l) / 2;
            let time_needed = Solution::time(&piles, m as f64);
            if time_needed <= h as i64 { // something to remember, Rust will not do automatic type converion
                res = res.min(m);
                r = m - 1;
            } else {
                l = m + 1;
            }
        }

        res as i32
    }

    fn time(piles: &Vec<i32>, h: f64) -> i64 {
        // want ceil(piles[i] / h) summed together
        piles.iter()
            .map(|&pile| (pile as f64 / h).ceil() as i64)
            .sum()
    }
}

fn main() {
    let piles = vec![805306368,805306368,805306368];
    println!("{}", Solution::min_eating_speed(piles, 1000000000));
}